### PR TITLE
Bugfix/rrfs fixes 1

### PR DESCRIPTION
**Pull request by MarcelCaron-NOAA**
<b>Note to developers: You must use this PR template!</b>

## Description of Changes

> Please include a summary of the changes and the related GitHub issue(s). Please also include relevant motivation and context.

Addresses most of the remaining fixes and additions needed for EVSv2.0, specifically those relevant to any EVS-RRFS jobs:
- SNOD and ASNOW fields are now verified independently from one another
- The CAPE verification configuration was fixed so that RRFS is verified against the same matched pairs as are other CAMs
- Added flexibility to verification against MRMS precip to reduce the impacts of missing MRMS files
- Added an NCO bugzilla fix, which runs child processes in separate directories before merging the output
- Various resource adjustments, and matching dev with ecf drivers

Other changes:
- Fixed mail message logic for missing RRFS files
- Updated various plots configurations, adding RRFS where the model was missing
- Changed cpreq to cp -rv where needed

## Developer Questions and Checklist
* Is this a high priorty PR? If so, why and is there a date it needs to be merged by?
No

* Do you have any planned upcoming annual leave/PTO?
No. 

* Are there any changes needed for when the jobs are supposed to run?
No.
  
- [x] The code changes follow [NCO's EE2 Standards](https://www.nco.ncep.noaa.gov/idsb/implementation_standards/ImplementationStandards.v11.0.0.pdf).
- [x] Developer's name is removed throughout the code and have used `${USER}` where necessary throughout the code.
- [x] References the feature branch for `HOMEevs` are removed from the code.
- [x] J-Job environment variables, COMIN and COMOUT directories, and output follow what has been [defined](https://docs.google.com/document/d/1JWg_4q80aYmmAoD21GFjp9R9y5-3w7WGM3-0HJk0Pjs/edit#heading=h.7ysbr191vzu4) for EVS.
- [x] Jobs over 15 minutes in runtime have restart capability.
- [x] If applicable, changes in the `dev/drivers/scripts` or `dev/modulefiles` have been made in the corresponding `ecf/scripts` and `ecf/defs/evs-nco.def`? 
- [x] Jobs contain the approriate file checking and don't run METplus for any missing data.
- [x] Code is using METplus wrappers structure and not calling MET executables directly.
- [x] Log is free of any ERRORs or WARNINGs.

## Testing Instructions

> Please include testing instructions for the PR assignee. Include all relevant input datasets needed to run the tests.

### (1) Set up jobs
• symlink the EVS_fix directory locally as "fix"

**In all driver scripts...**
• ... point all exports of "HOMEevs" to your EVS test directory
• ... change COMIN to my test directory (`/lfs/h2/emc/vpppg/noscrub/marcel.caron/test/evs_release_rrfsv1.0/$NET/$evs_ver_2d`)
• ... change COMOUT to your test COMOUT directory
• ... change DATAROOT to your test DATA root directory
• ... add the following line to developer settings:
```
export COMINrrfs=/lfs/h2/emc/ptmp/emc.lam/rrfs/${rrfs_ver}/prod
```

**In stats and plots driver scripts...**
• ... set SENDMAIL to `"NO"` (or adjust MAILTO)

### (2) I recommend testing the following jobs:
```
jevs_cam_rrfs_precip_prep
```
```
jevs_cam_rrfs_grid2obs_stats
jevs_cam_rrfs_precip_stats
jevs_cam_rrfs_radar_stats
jevs_cam_rrfs_severe_stats
jevs_cam_rrfs_snowfall_stats
```
```
jevs_cam_firewxnest_grid2obs_plots
jevs_cam_precip_plots_last90days
jevs_cam_snowfall_plots
```
[_Total: 1 prep job; 5 stats jobs; 3 plots jobs_]

### (3) Running jobs
**Most jobs need to be fed "vhr":**
• Apart from the exceptions listed below, all driver scripts can be submitted using `qsub -v vhr=00 $driver`
• For jevs_cam_rrfs_grid2obs_stats, use `qsub -v vhr=02 $driver`
• For jevs_cam_rrfs_severe_stats, use `qsub -v vhr=08 $driver`
• For jevs_cam_rrfs_precip_stats, use `qsub -v vhr=21,VDATE=$PDYm3 $driver` where `$PDYm3` is three days before present
• For jevs_cam_rrfs_precip_plots, use `qsub -v vhr=00,VDATE=$PDYm3 $driver` where `$PDYm3` is three days before present

**When to submit jobs:**
• all prep jobs: on the valid day anytime after the "vhr" hour fed to the job scheduler.
• severe stats jobs: on the valid day anytime after the "vhr" hour fed to the job scheduler.
• all other jobs: anytime.

### (4) Verifying jobs
Log files should be checked by the developer for the following keywords:
```
check="FATAL\|WARNING\|error\|Killed\|Cgroup\|argument expected\|No such file\|cannot\|failed\|unexpected\|exceeded"
grep "$check" $logfile
```
Additionally, output statistics, plot tar balls, and other major files should be compared to relevant archives (EVS parallel, RRFS routine verification, etc.).

### CODE REVIEW

**Comment by malloryprow** [comment]
@MarcelCaron-NOAA I see this is still in draft. Let me know when it is good to start code review and testing.

**Comment by MarcelCaron-NOAA** [comment]
@malloryprow Yes thanks for checking.  I want to get my notes in here while I finish testing the affected jobs.  I still need to tune the resource requirements for one job but should be able to submit soon.

**Comment by MarcelCaron-NOAA** [comment]
@malloryprow all set for code review and testing 👍

File: dev/drivers/scripts/plots/cam/jevs_cam_radar_nbrcnt_plots_last31days.sh [hunk 1]
```diff
@@ -4,7 +4,7 @@
 #PBS -q dev
 #PBS -A VERF-DEV
 #PBS -l walltime=1:50:00
-#PBS -l place=vscatter:exclhost,select=1:ncpus=128
+#PBS -l place=vscatter:exclhost,select=1:ncpus=64
```
Comment (malloryprow) [inline dev/drivers/scripts/plots/cam/jevs_cam_radar_nbrcnt_plots_last31days.sh]: Need to specify a `mem` here and its accompanying ecf script.

File: dev/drivers/scripts/plots/cam/jevs_cam_radar_nbrcnt_plots_last90days.sh [hunk 2]
```diff
@@ -4,7 +4,7 @@
 #PBS -q dev
 #PBS -A VERF-DEV
 #PBS -l walltime=1:50:00
-#PBS -l place=vscatter:exclhost,select=1:ncpus=128
+#PBS -l place=vscatter:exclhost,select=1:ncpus=64
```
Comment (malloryprow) [inline dev/drivers/scripts/plots/cam/jevs_cam_radar_nbrcnt_plots_last90days.sh]: Need to specify a `mem` here and its accompanying ecf script.

File: dev/drivers/scripts/plots/cam/jevs_cam_radar_nbrctc_plots_last31days.sh [hunk 3]
```diff
@@ -4,7 +4,7 @@
 #PBS -q dev
 #PBS -A VERF-DEV
 #PBS -l walltime=1:50:00
-#PBS -l place=vscatter:exclhost,select=1:ncpus=128
+#PBS -l place=vscatter:exclhost,select=1:ncpus=64
```
Comment (malloryprow) [inline dev/drivers/scripts/plots/cam/jevs_cam_radar_nbrctc_plots_last31days.sh]: Need to specify a `mem` here and its accompanying ecf script.

File: dev/drivers/scripts/plots/cam/jevs_cam_radar_nbrctc_plots_last90days.sh [hunk 4]
```diff
@@ -4,7 +4,7 @@
 #PBS -q dev
 #PBS -A VERF-DEV
 #PBS -l walltime=1:50:00
-#PBS -l place=vscatter:exclhost,select=1:ncpus=128
+#PBS -l place=vscatter:exclhost,select=1:ncpus=64
```
Comment (malloryprow) [inline dev/drivers/scripts/plots/cam/jevs_cam_radar_nbrctc_plots_last90days.sh]: Need to specify a `mem` here and its accompanying ecf script.

**Comment by malloryprow** [comment]
Left some comments for some simple fixes. Can you do a sync fork when you get a chance too?

**Comment by malloryprow** [comment]
Will this code correctly work with beta5? Does it have the fixes for unit conversion for plots?

**Comment by MarcelCaron-NOAA** [comment]
@malloryprow all fixes you mentioned are in.  

Also yes this should work with beta5, and yes the unit conversion fix is in!

**Comment by malloryprow** [comment]
### prep

Ran with `export COMIN=/lfs/h2/emc/vpppg/noscrub/marcel.caron/test/evs_release_rrfsv1.0/$NET/$evs_ver_2d`, `export COMINrrfs=/lfs/h2/emc/ptmp/emc.lam/rrfs/${rrfs_ver}/prod`. Submitted with `qsub -v vhr=00`.

### 1. jevs_cam_rrfs_precip_prep.sh
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/prep/cam/jevs_cam_rrfs_precip_prep.o208210318
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_rrfs_precip_prep.208210318.dbqs01
**COMOUT:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/evs/v2.0/prep/cam

**Comment by MarcelCaron-NOAA** [comment]
✔️  **jevs_cam_rrfs_precip_prep** completed successfully and output look good.  MPMD and MRMS precip flexibility fixes worked as expected.

👍 no more concerns from me about **prep**

**Comment by malloryprow** [comment]
I can comment again on PRs so reiterating what I sent in an email.

### stats
COMOUT for all is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/evs/v2.0/stats/cam.

### 1. jevs_cam_rrfs_grid2obs_stats.sh (vhr=02)
**Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/stats/cam/jevs_cam_rrfs_grid2obs_stats.o208219187
**DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_rrfs_grid2obs_stats.208219187.dbqs01

### 2. jevs_cam_rrfs_precip_stats (vhr=21,VDATE=20241130)
**Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/stats/cam/jevs_cam_rrfs_precip_stats.o208219579
**DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_rrfs_precip_stats.208219579.dbqs01

### 3. jevs_cam_rrfs_radar_stats (vhr=00)
**Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/stats/cam/jevs_cam_rrfs_radar_stats.o208219724
**DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_rrfs_radar_stats.208219724.dbqs01

### 4. jevs_cam_rrfs_severe_stats (vhr=08)
**Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/stats/cam/jevs_cam_rrfs_severe_stats.o208219812
**DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_rrfs_severe_stats.208219812.dbqs01

### 5. jevs_cam_rrfs_snowfall_stats (vhr=00)
**Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/stats/cam/jevs_cam_rrfs_snowfall_stats.o208219895
**DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_rrfs_snowfall_stats.208219895.dbqs01

**Comment by malloryprow** [comment]
jevs_cam_rrfs_severe_stats (and its associated ecf) are requesting the whole node because it is requesting 500GB of memory. It is only using 0.076368 GB (76368kb).

Can we change so it is `place=shared,select=1:ncpus=1:mem=10MB`?

**Comment by MarcelCaron-NOAA** [comment]
Oh oops!  That is carryover from other severe jobs that were fixed.  yes let me change that now.

**Comment by MarcelCaron-NOAA** [comment]
Done.  It is a quick job (jevs_cam_rrfs_severe_stats) so feel free to rerun with the resource adjustments!

**Comment by malloryprow** [comment]
### stats
Reran radar and severe stats just to be sure. COMOUT for all is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/evs/v2.0/stats/cam.

### 1.  jevs_cam_rrfs_radar_stats (vhr=00)
**Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/stats/cam/jevs_cam_rrfs_radar_stats.o208298210
**DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_rrfs_radar_stats.208298210.dbqs01

### 2. jevs_cam_rrfs_severe_stats (vhr=08)
**Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/stats/cam/jevs_cam_rrfs_severe_stats.o208298483
**DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_rrfs_severe_stats.208298483.dbqs01

**Comment by MarcelCaron-NOAA** [comment]
The following completed cleanly and output look normal:
✔️ jevs_cam_rrfs_grid2obs_stats
✔️ jevs_cam_rrfs_precip_stats
✔️ jevs_cam_rrfs_snowfall_stats

The following completed with walltime and/or memory exceedence:
🏁 jevs_cam_rrfs_severe_stats (walltime and mem)
🏁 jevs_cam_rrfs_radar_stats (mem)

We should rerun the radar and severe jobs.  Let me adjust and push the changes!

**Comment by MarcelCaron-NOAA** [comment]
@malloryprow I tested the two jobs with updated resource allocations and they finished successfully.  I think we can re-test those two jobs here.

**Comment by malloryprow** [comment]
### stats
COMOUT for all is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/evs/v2.0/stats/cam. I cleared out rrfs.20241203 and atmos.20241203 to avoid restart.

### 1.  jevs_cam_rrfs_radar_stats (vhr=00)
**Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/stats/cam/jevs_cam_rrfs_radar_stats.o208311944
**DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_rrfs_radar_stats.208311944.dbqs01

### 2. jevs_cam_rrfs_severe_stats (vhr=08)
**Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/stats/cam/jevs_cam_rrfs_severe_stats.o208311919
**DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_rrfs_severe_stats.208311919.dbqs01

**Comment by MarcelCaron-NOAA** [comment]
The two remaining stats jobs completed cleanly and output look normal:
✔️ jevs_cam_rrfs_severe_stats
✔️ jevs_cam_rrfs_radar_stats

👍 no other concerns from me about **stats**

**Comment by malloryprow** [comment]
For` jevs_cam_firewxnest_grid2obs_plots`, it runs `scripts/plots/cam/exevs_cam_firewxnest_grid2obs_plots.sh`? If so we can change the resources to `select=1:ncpus=1`.

**Comment by MarcelCaron-NOAA** [comment]
@malloryprow Ah right.  Fixed!

**Comment by malloryprow** [comment]
### plots

COMOUT is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/evs/v2.0/plots/cam

### 1. jevs_cam_firewxnest_grid2obs_plots (vhr=00)
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/plots/cam/
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_firewxnest_grid2obs_plots.208317981.dbqs01

### 2. jevs_cam_precip_plots_last90days (vhr=00,VDATE=20241201)
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/plots/cam/
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_precip_plots_last90days.208317113.dbqs01

### 3. jevs_cam_snowfall_plots (vhr=00)
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/plots/cam/jevs_cam_snowfall_plots.o208317358
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_snowfall_plots.208317358.dbqs01

**Comment by MarcelCaron-NOAA** [comment]
The following completed cleanly and output look normal:
✔️ jevs_cam_snowfall_plots

The following completed with a minor issue:
⚠️ jevs_cam_firewxnest_grid2obs_plots

Still running:
⏳ jevs_cam_precip_plots_last90days

@malloryprow can you rerun **jevs_cam_firewxnest_grid2obs_plots** with the fix?  It ran nearly successfully but the nam_firewxnest stats were not included.  Should work now.

**Comment by MarcelCaron-NOAA** [comment]
✔️ **jevs_cam_precip_plots_last90days** completed cleanly and output look normal.

**Comment by malloryprow** [comment]
### plots

Ah, not sure how I missed your request to rerun the firewx_nest. It is running now. COMOUT is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/evs/v2.0/plots/cam

### 1. jevs_cam_firewxnest_grid2obs_plots (vhr=00)
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/plots/cam/jevs_cam_firewxnest_grid2obs_plots.o208375394
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_firewxnest_grid2obs_plots.208375394.dbqs01

**Comment by MarcelCaron-NOAA** [comment]
Thanks @malloryprow!  The recent job ran nearly fine, but printed the following:

```
12/05/2024 12:32:50  M    ncpus 1.6 exceeded limit 1 (burst)
```

I suspect there's some multithreading or parallelization happening somewhere.  Mind if I increase `ncpus=2` and we rerun?

**Comment by malloryprow** [comment]
Interesting...I can change on my end and re-run.

**Comment by malloryprow** [comment]
### plots

Job submitted and in the queue. COMOUT is /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/evs/v2.0/plots/cam (moved the previous run to /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/evs/v2.0/plots/cam/save.atmos.20241204). 

### 1. jevs_cam_firewxnest_grid2obs_plots (vhr=00 w/ ncpus=2)
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr613/EVS/dev/drivers/scripts/plots/cam/jevs_cam_firewxnest_grid2obs_plots.o208400524
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_cam_firewxnest_grid2obs_plots.208400524.dbqs01

**Comment by MarcelCaron-NOAA** [comment]
✔️ **jevs_cam_firewxnest_grid2obs_plots** completed cleanly with the updated resources and output look normal

👍 No other concerns from me about the affected jobs in this PR

**Comment by malloryprow** [comment]
Wonder where those resources are coming from, odd but glad it is sorted out.

**Comment by malloryprow** [comment]
I asked some of these before but the RRFS code has all these things right? Before I give it approval. ✅ 

1. CAPE configuration changes
2. Beta5 changes
3. Unit Conversion fix for plots
4. MPMD individual working directories

**Comment by MarcelCaron-NOAA** [comment]
> Wonder where those resources are coming from, odd but glad it is sorted out.

I'm not sure, maybe @PerryShafran-NOAA has an idea?

**Comment by MarcelCaron-NOAA** [comment]
> I asked some of these before but the RRFS code has all these things right? Before I give it approval. ✅
> 
> 1. CAPE configuration changes
> 2. Beta5 changes
> 3. Unit Conversion fix for plots
> 4. MPMD individual working directories

1) CAPE configuration changes were made in the current PR [(relevant commit)](https://github.com/NOAA-EMC/EVS/pull/613/commits/4da41b93ad63393c480c4788260cf8583cb02f4b), and tested successfully
2) Beta5 changes were brought in [indirectly via feature/rrfs_refs_v1](https://github.com/NOAA-EMC/EVS/commit/aad674e8c95c6e55781f4232493030747328b440); confirmed in ush/cam/plot_util.py
3) Unit conversion fix was brought in [indirectly via feature/rrfs_refs_v1](https://github.com/NOAA-EMC/EVS/commit/4d49f827d46badd7a404f63902da56e35261d0b6); confirmed in ush/cam/plot_util.py
4) The MPMD bugzilla fix was made in the current PR [(relevant commit)](https://github.com/NOAA-EMC/EVS/pull/613/commits/687db3d62b412ab307873ab4c417453edfba88e2), and tested successfully

**Comment by PerryShafran-NOAA** [comment]
I don't recall why those resources were originally selected for the nam_firewxnest, and I don't know if the rrfs_firewxnest is much different to be honest.  In any case, good that they are sorted out.

**Comment by malloryprow** [comment]
> > I asked some of these before but the RRFS code has all these things right? Before I give it approval. ✅
> > 
> > 1. CAPE configuration changes
> > 2. Beta5 changes
> > 3. Unit Conversion fix for plots
> > 4. MPMD individual working directories
> 
> 1. CAPE configuration changes were made in the current PR [(relevant commit)](https://github.com/NOAA-EMC/EVS/pull/613/commits/4da41b93ad63393c480c4788260cf8583cb02f4b), and tested successfully
> 2. Beta5 changes were brought in [indirectly via feature/rrfs_refs_v1](https://github.com/NOAA-EMC/EVS/commit/aad674e8c95c6e55781f4232493030747328b440); confirmed in ush/cam/plot_util.py
> 3. Unit conversion fix was brought in [indirectly via feature/rrfs_refs_v1](https://github.com/NOAA-EMC/EVS/commit/4d49f827d46badd7a404f63902da56e35261d0b6); confirmed in ush/cam/plot_util.py
> 4. The MPMD bugzilla fix was made in the current PR [(relevant commit)](https://github.com/NOAA-EMC/EVS/pull/613/commits/687db3d62b412ab307873ab4c417453edfba88e2), and tested successfully

Perfect! This will be ready for operations when the time comes :)

**Review by malloryprow — APPROVED** [review]
Code changes are good and testing successful.

**Review by AndrewBenjamin-NOAA — APPROVED** [review]
These changes look good, and I approve this PR. Thanks for getting this into the RRFS feature branch

**Pull request merged**